### ISSUE TITLE
Benchmark different CNI pairings 

### ISSUE THREAD

**Issue by Kale98** [testground/testground#1411]
## What defines this endeavour to be complete?

<!--Create a check list of what you expect to see as part of delivering on the ask--

- [ ] Item 1
- [ ] Item 2
- [ ] ...

**Comment by laurentsenta** [comment]
Hey team, following the discussion from yesterday, could you share:

- the list of CNI pairs we are going to benchmark,
- the code used to benchmark & the results,
- please feel free to create sub-tasks per CNI if it's too much content for a single issue.

**Comment by dektech** [comment]
Currently we have benchmarked `calico/weave`, and now running nodegroup scaling to see how it performs after that (#1410).

Then, as discussed yesterday, we will work on `AWS VPC CNI/weave` to see if we can match the performance obtained from the above combination, which is 250 storm instances on 2 c5.4xlarge nodes.

The code used to test is a `bash` script that accepts number of runs and number of instances for the run, then measures how much total time has passed, and of course the number of successful/failed tests.
But at the core of it is the following command:

```
testground run single --plan=benchmarks --testcase=storm --builder=docker:go --runner=cluster:k8s --instances=250
```

Entire script:

```
#!/bin/bash
test_start_time="$(date -u +%s)"
# Run many testground runs, one by one, and print stats
#
# Usage:
# ./perf_test.sh <run_count> <plan> <case> <instance_count> <builder> <runner>
# e.g.
# ./perf_test.sh 50 benchmarks storm 5 docker:go local:docker

function assert_run_output_is_correct {

    outcome=$(awk '/run finished with outcome = / {if ($10=="failure"){print 1} else {print 0} fi}' $1)
    # $outcome is already numeric? Cannot be used as a return value for some reason
    if [ $outcome -eq 1 ] ; then
        return 1
    else
        return 0
    fi
}

function perform_runs {
    runs=$1
    plan=$2
    case=$3
    instances=$4
    builder=$5
    runner=$6

    # total number of failed test runs
    failed=0

    for i in `seq 1 $1`; do
        # run single, wait for it to finish and output to out.txt
        testground run single --plan $plan --testcase $case \
            --builder $5 \
            --runner $6 \
            --instances $instances --wait | tee out.txt

        # check outcome
        assert_run_output_is_correct out.txt
        outcome=$?
        failed=`expr $failed + $outcome`
        echo "End #$i Outcome $outcome. Failed so far: $failed"
    done

    test_end_time="$(date -u +%s)"
    total_test_time="$(($test_end_time-$test_start_time))"

    # calculate stats and exit
    success=$(($runs - $failed))
    success_rate=$(bc <<< "scale=2; ($success/$runs)*100")
    echo "==========================="
    echo "Finished $1 runs"
    echo "Total succeeded: $(($success))"
    echo "Total failed: $failed"
    echo "Success rate: $success_rate%"
    echo "Total test time: $total_test_time seconds"
    echo "==========================="
}

# sanity check: arguments
if [ $# -lt 6 ]; then
    echo "No argument supplied: supply a number indicating number of runs, test plan, test case, instances #, builder and runner"
    exit 1
fi

perform_runs $1 $2 $3 $4 $5 $6
```

Output at the end of a run:

```
===========================
Finished 20 runs
Total succeeded: 20
Total failed: 0
Success rate: 100.00%
Total test time: 7169 seconds
===========================
```

**Comment by laurentsenta** [comment]
Great, thanks for sharing,

Can you describe the list of `(number of hosts, plan, case, numbers of runs,  number of instances)` we're going to run?
I guess it's already scripted somewhere,

Also, this is running 1 test N times,
A realistic load for the TaaS cluster might be multiple jobs scheduled and running simultaneously.

- Is this something that might impact stability/perf and our choice of CNI?
- Are we planning on benchmarking this?

To illustrate my thinking, do we need a case that would look like this:

```sh

    for i in `seq 1 $1`; do
        # mixer test case
        $(testground run single storm --instances small --wait) &
        $(testground run single storm --instances large --wait) &
        $(sleep 1min && testground run single storm --instances medium --wait) &
        $(sleep 20s && testground run single ping--instances small --wait) &
        $(sleep 1min && testground run single ping--instances large --wait) &
        $(sleep 2min && testground run single ping--instances small --wait) &
        $(sleep 2min && testground run single ping--instances medium --wait) &
        wait;
        # mixer test case

        ...assert, etc.
```

**Comment by dektech** [comment]
As explained in the previous post, this command is the core of the script, and I believe that answers the first question:
```testground run single --plan=benchmarks --testcase=storm --builder=docker:go --runner=cluster:k8s --instances=250```

And you are correct, this is running that test N times. Next steps are pretty much as you describe it - running simultaneous tests and measure performance and stability.

**Comment by dektech** [comment]
We have ran several iterations of simultaneous tests using different number of processes and instances (within hardware limits, of course), and we always get the same result - 2 simultaneous processes will run, the 3rd will wait for any of the other two to finish before running.
It will be scheduled, but will not run until previous runs are completed. So, for example, it looks like this:

We schedule 3 separate runs at the same time (let's say there is 1 second difference between them) - `R1`, `R2`, and `R3` - and have it run 2 times each ( 2x `R1`, 2x `R2`, and 2x `R3`)

`R1` and `R2` start, while `R3` is only scheduled
`R1` finishes and immediately triggers the 2nd iteration of `R1`
`R2` finishes, and while doing the cleanup, `R3` starts
now we have `R1` and `R3` running, while `R2` is waiting to start
as soon as `R1` finishes, 2nd iteration of `R2` starts
since `R1` finished both its runs (this test included only 2 iterations), now `R2` and `R3` will finish their 2 iterations without waiting

The results look like this, and we can clearly see that R1 finished much before R2 and R3 simply because it was scheduled first:

```
===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 448 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 652 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 653 seconds
===========================
```

Other than that, there are no issues with running simultaneous tests. The tasks will finish approximately after the same duration compared to running single tasks.

**Comment by KatarinaKeti** [comment]
CALICO/WEAVE

==
2 x 10 storm instances; 3 parallel processes
==

1st try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 266 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 269 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 177 seconds
===========================

2nd try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 252 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 176 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 250 seconds
===========================

3rd try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 238 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 246 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 184 seconds
===========================

4th try with 4 parallel processes

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 262 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 266 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 338 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 347 seconds
===========================

==
2 x 50 storm instances; 3 parallel processes
==

1st try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 453 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 660 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 656 seconds
===========================

2nd try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 448 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 649 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 651 seconds
===========================

3rd try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 448 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 652 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 653 seconds
===========================

==
2 x 100 storm instances; 2 parallel processes
==

1st try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 698 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 696 seconds
===========================

2nd try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 723 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 727 seconds
===========================

############################
############################
############################

AWS VPC CNI / WEAVE

3x10 pods, not multiple runs

===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 221 seconds
===========================

3x50 pods, not multiple runs

===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 357 seconds
===========================

3x100 pods, not multiple runs

===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 517 seconds
===========================

20x100 pods, not multiple runs

===========================
Finished 20 runs
Total succeeded: 20
Total failed: 0
Success rate: 100.00%
Total test time: 3492 seconds
===========================

1x250 pods

===========================
Finished 1 runs
Total succeeded: 1
Total failed: 0
Success rate: 100.00%
Total test time: 302 seconds
===========================

==
3 x 40 storm instances; 3 parallel processes
==

===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 579 seconds
===========================

===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 582 seconds
===========================

===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 678 seconds
===========================

==
3 x 100 storm instances; 3 parallel processes
==

===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 736 seconds
===========================

===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 991 seconds
===========================

===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 1155 seconds
===========================

**Comment by brdji** [comment]
Here's the performance of simultaneous test runs on the Calico (primary) + Weave (secondary) CNI:

==
2 x 10 storm instances; 3 parallel processes
==

1st try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 266 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 269 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 177 seconds
===========================

2nd try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 252 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 176 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 250 seconds
===========================

3rd try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 238 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 246 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 184 seconds
===========================

4th try with 4 parallel processes

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 262 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 266 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 338 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 347 seconds
===========================

==
2 x 50 storm instances; 3 parallel processes
==

1st try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 453 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 660 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 656 seconds
===========================

2nd try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 448 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 649 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 651 seconds
===========================

3rd try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 448 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 652 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 653 seconds
===========================

==
2 x 100 storm instances; 2 parallel processes
==

1st try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 698 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 696 seconds
===========================

2nd try

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 723 seconds
===========================

===========================
Finished 2 runs
Total succeeded: 2
Total failed: 0
Success rate: 100.00%
Total test time: 727 seconds
===========================

############################
############################

**Comment by brdji** [comment]
Here's the performance of AWS VPC CNI (Primary) + Weave (secondary) CNI cluster:

3x10 pods, not multiple runs

```
===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 221 seconds
===========================
```
3x50 pods, not multiple runs
```
===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 357 seconds
===========================
```
3x100 pods, not multiple runs
```
===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 517 seconds
===========================
```
20x100 pods, not multiple runs
```
===========================
Finished 20 runs
Total succeeded: 20
Total failed: 0
Success rate: 100.00%
Total test time: 3492 seconds
===========================
```
1x250 pods
```
===========================
Finished 1 runs
Total succeeded: 1
Total failed: 0
Success rate: 100.00%
Total test time: 302 seconds
===========================
```
==
3 x 40 storm instances; 3 parallel processes
==
```
===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 579 seconds
===========================

===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 582 seconds
===========================

===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 678 seconds
===========================
```
==
3 x 100 storm instances; 3 parallel processes
==
```
===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 736 seconds
===========================

===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 991 seconds
===========================

===========================
Finished 3 runs
Total succeeded: 3
Total failed: 0
Success rate: 100.00%
Total test time: 1155 seconds
===========================
```

**Comment by BigLep** [comment]
@brdji : if I'm understanding this right, it looks like we did comparisons between:
- Calico (primary) + Weave (secondary) CNI:
- AWS VPC CNI (Primary) + Weave (secondary) CNI cluster

What are the conclusions from the benchmark?  What decisions are we making as a result?

**Comment by AbominableSnowman730** [comment]
TL;DR: We should stick with AWS VPC as the primary CNI plugin/control network, since it provides the fastest data path, and is AWS native. We have to stick with Weave as the secondary CNI plugin/data network, since it's the only plugin that provides necessary functionality for Testground to run properly.

A little longer answer would be that, practically, we could choose to run Testground on an EKS cluster that uses Calico (most likely any other actively developed CNI plugin) with slight performance loss (which shouldn't have a large impact on tests we want to run). This gives us confidence that we can substitute AWS VPC CNI with any other battle tested plugin, and wouldn't incur major penalties. On this particular occasion this goes along with what Calico is claiming that they utilize "smart routing" for connectivity between pods on different nodes, even in different subnets and AZs.

On the other hand, using AWS VPC as the default option for control network connectivity should be the path to take, mainly for two reasons:

- We don't need to worry about routing, or any other networking issues we might face. AWS is in charge of it, and we're directly utilizing their VPC to establish the primary connection. We're building on top of the L2 network connectivity established by AWS. It's going to be as fast as reliable as it gets.
-  We're delegating all the maintenance work to the AWS.  


As far as the secondary - data - network goes, we're still relying on Weave, since it is the only sound solution for manipulating container attachments/detachments from a network. This is one of the [original requirements](https://github.com/testground/docs/blob/master/misc/presentations/k8s-networkings.slide#L86).
On this note, Multus v4 should incorporate this functionality for any running plugin, but it's still not officially released, and the designated controller would be something that's developed by a 3rd party. This is something we should monitor closely, and try to incorporate when time comes.
